docs/test_plans/0003-Task_Throughput.md

```diff
@@ -0,0 +1,132 @@
+
+## Current State
+We're gonna use the perf cluster to run those tests, under the pulp-perf namespace.
+Each pulp-api pod should replicate the same resource configuration from pulp-stage 
+on stage cluster.
+
+For database, we're using the AWS RDS PostgreSQL 16, on a db.m7g.2xlarge instance class.
+In theory, the maximum number of connections is defined by `LEAST({DBInstanceClassMemory/9531392}, 5000)`
+as written [here](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/CHAP_Limits.html#RDS_Limits.MaxConnections)
+
+## Metrics to follow
+On database level:
+- Active Connections / Sessions
+- Connection wait time
+- Slow queries
+- CPU and Memory utilization
+
+Most of those metrics can be checked [here](https://us-east-1.console.aws.amazon.com/rds/home?region=us-east-1#database:id=pulp-prod;is-cluster=false)
+
+For the application, we need to follow the timeouts using the logs.
+- Tasks Completed per Second (TPS)
+- Task Success Rate
+- Task Latency/Completion Time
+- Error Rates
+- Queue Lengths
+You can start [here](https://grafana.app-sre.devshift.net/explore?schemaVersion=1&panes=%7B%22vse%22%3A%7B%22datasource%22%3A%22P1A97A9592CB7F392%22%2C%22queries%22%3A%5B%7B%22id%22%3A%22%22%2C%22region%22%3A%22us-east-1%22%2C%22namespace%22%3A%22%22%2C%22refId%22%3A%22A%22%2C%22queryMode%22%3A%22Logs%22%2C%22expression%22%3A%22fields+%40logStream%2C+%40message%2C++kubernetes.namespace_name+%7C+filter+%40logStream+like+%2Fpulp-stage_pulp-%28worker%7Capi%7Ccontent%29%2F%5Cn%5Cn%5Cn%5Cn%22%2C%22statsGroups%22%3A%5B%5D%2C%22datasource%22%3A%7B%22type%22%3A%22cloudwatch%22%2C%22uid%22%3A%22P1A97A9592CB7F392%22%7D%2C%22logGroups%22%3A%5B%7B%22arn%22%3A%22arn%3Aaws%3Alogs%3Aus-east-1%3A744086762512%3Alog-group%3Acrcs02ue1.pulp-stage%3A*%22%2C%22name%22%3A%22crcs02ue1.pulp-stage%22%2C%22accountId%22%3A%22744086762512%22%7D%5D%7D%5D%2C%22range%22%3A%7B%22from%22%3A%22now-30m%22%2C%22to%22%3A%22now%22%7D%7D%7D&orgId=1)
+
+## Test Plan
+
+1. Start the example script firing a number of requests to create an empty distribution.
+2. Using the logs, calculate the number of successful tasks, and the number of tasks completed per second.
+3. Calculate the Task Latency using a sample of executed tasks.
+
+
+## Results
+Date of the test: YYYY/mm/dd
+
+A possible template for the table could be:
+
+| Run  | Workers | Requests | Successful Tasks | Task Latency | Error Rates | Observations |
+|------|---------|----------|------------------|--------------|-------------|--------------|
+| 1    | 1       | 1000     | 990              | 10s          | 1/min       | `...`        | 
+| 2              | {{value}}| {{value}}        | `...`        | `...`       | `...`        |
+...
+
+
+**Key:**
+- Use `-` for unavailable metrics.
+
+## An example to calculate the number of tasks ingested
+```python
+import requests
+import time
+from concurrent.futures import ThreadPoolExecutor, TimeoutError
+
+def send_request_and_process_tasks(url):
```

Comment:
I think you want to basically use the code you used exactly from the ingestion test and pick a reasonable maximum of "ingestion API tasks" which from the previous data I read as N=2 as being a good choice. So in that scenario you'll have two pulpcore-api workers creating tasks like crazy, and the workers running in the background and picking them up. Just link to that code and call out how it's going to be run. You don't need to "rewrite" it here.

Reply:
The new code here you need to:
1. Wait for all tasks to complete (you don't need to calculate as it runs). Or just manually check it and run your calculation code afterwards.
2. Find the start time and end time of the test. I recommend finding the very first task created_at time as the start time, and the very latest end_time as the end time. Use the difference in those timestamps as the duration of the test (time_in_seconds_duration).
3. Compute the rate from the number of tasks / time_in_seconds_duration.